### ISSUE TITLE
`Literal` type is widened in a lambda expression when originating from iteration over `dict.items()`

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#3487]
Pyright widens `Literal` type when used in a lambda expression when the source of the `Literal` type is the key and value from a `dict.items()` expression.

Example code:

```py
from typing import Literal

def some_func(some_dict: dict[Literal["hello"], Literal[5]]):
    for hello, five in some_dict.items():
        reveal_type(hello)  # Type of "hello" is "Literal['hello']"
        lambda x: reveal_type(hello)  # Type of "hello" is "str"

        reveal_type(five)  # Type of "five" is "Literal[5]"
        lambda x: reveal_type(five)  # Type of "five" is "int"
```

~The `Literal` type does not get widened when iterating over either `dict.keys()` or `dict.values()` or when iterating over other iterable types as far as I've noticed, although I have not conducted an exhaustive search.~

Mypy in this case retains the `Literal` type, and I would expect pyright to do the same.

Tested on pyright version `1.1.247` and mypy version `0.950`

**Comment by kennipj** [comment]
Looks like my initial statement that this only affects `dict.items()` is not quite right, it seems to be more complicated than I thought.

See this example as well:

```py
from typing import Literal

def example_1(hello_list: list[Literal["hello"]]):
    for hello in hello_list:
        lambda x: reveal_type(hello) # Type of "hello" is "Literal['hello']"
        map(lambda x: reveal_type(hello), hello_list) # Type of "hello" is "Literal['hello']"
        lambda x: reveal_type(hello) # Type of "hello" is "str"
        reveal_type(hello) # Type of "hello" is "Literal['hello']"
```

**Comment by erictraut** [comment]
The inferred type of `hello` in this case is `str`. It is narrowed to `Literal["hello"]` upon assignment locally within the `some_func` execution context. The lambda captures the variable and has a distinct execution context. The narrowed type from the outer scope is lost. I think this is defensible behavior. I'll give further thought to whether we can safely retain the literal type in this case.

The reason this works with mypy is that it treats initial assignments as implicit type declarations, so `hello` is treated as though it has a declared type of `Literal["str"]` based on the assignment in the `for` statement. Confusing declared types with inferred types creates a bunch of problems in mypy, which is why pyright doesn't do it this way.

**Comment by erictraut** [comment]
If your intent is for `hello` to have a type of `Literal["hello"]`, you have the option of declaring its type explicitly.

```python
    hello: Literal["hello"]
    for hello, five in some_dict.items():
        ...
```

**Comment by kennipj** [comment]
I suppose the confusing part for me here, is that in some cases, like with `example_1`, the lambda expression before the `map(...)` expression the type is still `Literal["hello"]`. However, when iterating over `dict.items()` this is not the case, and it's not obvious to me why `dict.items()` would behave differently.

**Comment by erictraut** [comment]
Yeah, I could see what that's confusing. There is a bug in `example_1`. In this example, `hello` should not evaluate to `Literal['hello']`. It should evaluate to `str`. If you insert a dummy assignment between the `for` statement and the `lambda` statement (for example, `dummy=1`), you will see that the behavior changes, and `hello` is then (correctly) evaluated to `str`.

Here's what's going on here. The variable `hello` doesn't have a declared type, so its type is inferred based on its assignments. In both the case of `some_func` and `example_1`, the inferred type is `str`. However, in both cases, the assignment also locally narrows the type to `Literal['hello']`.

When the lambda is evaluated, pyright needs to determine whether to use the narrowed type of the captured variable or whether to use the unnarrowed type. In the general case, it is not safe to use the narrowed type because a type checker doesn't know when the lambda will be called, and the type could change after the variable is captured. Pyright allows the narrowed type to be used for the captured variable only if it can prove that there's no code path where the variable is assigned a new value after the lambda definition. In both `some_func` and `example_1`, the lambda is located within a while loop, and `hello` is reassigned every time through the loop. That means the captured variable cannot safely take on the narrowed type.

I've fixed the bug so you will see consistent behavior in `some_func` and `example_1`. In both cases, the revealed type of `hello` within the lambda should be `str`.

If your intent is for `hello` to be `Literal['hello']` within the lambda, you can provide a literal type declaration.

```python
def example_1(hello_list: list[Literal["hello"]]):
    hello: Literal["hello"]
    for hello in hello_list:
        lambda x: reveal_type(hello)
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.248, which I just published. It will also be included in the next release of pylance.
